### DISCUSSION TITLE
Any examples of headless-platform for ui-testing using Nunit framework?

### DISCUSSION THREAD

**Discussion by tyHdev** [AvaloniaUI/Avalonia#7655] [Q&A] [answered]
I am attempting to configure automated UI tests using Nunit and the Avalonia headless platform but having some difficulties. I am new to problems pertaining to concurrency, which I think is what is at play here, although my lack of complete understanding about the inter-workings of both XUnit and NUnit may be contributing as well.

I used both of the following as inspiration:
[Camelot source](https://github.com/IngvarX/Camelot/blob/cd7689af51d3965842b4c9e4924dc101f8b88bec/tests/Camelot.Ui.Tests/AvaloniaUiTestFramework.cs#L79) (huge shout out to the dev by the way)
[Avalonia Headless Unit Tests](https://github.com/kekekeks/Avalonia-unit-testing-with-headless-platform/blob/master/HeadlessTests/AvaloniaTestFramework.cs)

I took the code in SetupSyncContext() and tried adding it to an NUnit Setup Fixture:
```csharp
    [OneTimeSetUp]
    public void Setup()
    {
        SpawnAppUnderTestThread();
    }

    private void SpawnAppUnderTestThread()
    {
        var tcs = new TaskCompletionSource<SynchronizationContext>();
        var thread = new Thread(() =>
        {
            try
            {
                AvaloniaAppUnderTest
                    .BuildAvaloniaApp()
                    .AfterSetup(_ =>
                    {
                        tcs.SetResult(SynchronizationContext.Current);
                    })
                    .StartWithClassicDesktopLifetime(new string[0]);
            }
            catch (Exception e)
            {
                tcs.SetException(e);
            }
        })
        {
            IsBackground = true
        };

        thread.Start();

        SynchronizationContext.SetSynchronizationContext(tcs.Task.Result);
    }
```

When I try to access the application MainWindow from my test method, I get a "System.InvalidOperationException : Call from invalid thread" exception.

Maybe this code needs to be executed via framework extension of some kind, rather than the Setup method of a SetupFixture?

If sharing more code would be useful I can do that, but I thought first I would try to keep things simple and ask: Has anyone configured automated UI tests of their Avalonia application using the NUnit framework, and would you mind sharing a link to that source? 

Much appreciated in advance, thanks all!

**Comment by danipen** [comment, marked as answer]
https://github.com/kekekeks/Avalonia-unit-testing-with-headless-platform

**Reply by maxkatz6** [reply]
Up to date: https://docs.avaloniaui.net/docs/next/concepts/headless/
